source/common/router/scoped_rds.cc

```diff
@@ -0,0 +1,269 @@
+  std::vector<envoy::api::v2::ScopedRouteConfiguration> scoped_routes;
+  for (const auto& resource_any : resources) {
+    scoped_routes.emplace_back(
+        MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(resource_any));
+  }
+
+  std::unordered_set<std::string> resource_names;
+  for (const auto& scoped_route : scoped_routes) {
+    if (!resource_names.insert(scoped_route.name()).second) {
+      throw EnvoyException(
+          fmt::format("duplicate scoped route configuration {} found", scoped_route.name()));
+    }
+  }
+  for (const auto& scoped_route : scoped_routes) {
+    MessageUtil::validate(scoped_route);
+  }
+
+  // TODO(AndresGuedez): refactor such that it can be shared with other delta APIs (e.g., CDS).
+  std::vector<std::string> exception_msgs;
+  // We need to keep track of which scoped routes we might need to remove.
+  ScopedConfigManager::ScopedRouteMap scoped_routes_to_remove =
+      scoped_config_manager_.scopedRouteMap();
+  for (auto& scoped_route : scoped_routes) {
+    const std::string& scoped_route_name = scoped_route.name();
+    try {
+      scoped_routes_to_remove.erase(scoped_route_name);
+      ScopedRouteInfoConstSharedPtr scoped_route_info =
+          scoped_config_manager_.addOrUpdateRoutingScope(scoped_route, version_info);
+      if (scoped_route_info == nullptr) {
+        throw EnvoyException(
+            fmt::format("failed to create/update global routing scope {}", scoped_route_name));
+      }
+      ENVOY_LOG(debug, "srds: add/update scoped_route '{}'", scoped_route_name);
+      applyDeltaConfigUpdate(
+          [scoped_route_info](const ConfigProvider::ConfigConstSharedPtr& config) {
+            auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
+                static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
+            thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
+          });
+    } catch (const EnvoyException& ex) {
+      exception_msgs.push_back(fmt::format("{}: {}", scoped_route_name, ex.what()));
+    }
+  }
+
+  for (const auto& scoped_route : scoped_routes_to_remove) {
+    const std::string scoped_route_name = scoped_route.first;
+    ENVOY_LOG(debug, "srds: remove scoped route '{}'", scoped_route_name);
+    scoped_config_manager_.removeRoutingScope(scoped_route_name);
+    applyDeltaConfigUpdate([scoped_route_name](const ConfigProvider::ConfigConstSharedPtr& config) {
+      auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
+          static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
+      thread_local_scoped_config->removeRoutingScope(scoped_route_name);
+    });
+  }
+
+  ConfigSubscriptionCommonBase::onConfigUpdate();
+  setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(fmt::format("Error adding/updating scoped route(s) {}",
```

Comment:
Ditto

Reply:
This code has also been removed. See my response to the previous comment.